File: src/algorithms/cpp/binary-search.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

void emit(const string& type, const string& payload) {
    cout << "{\"type\":\"" << type << "\", " << payload << "}" << endl;
}

int binary_search(const vector<int>& arr, int target) {
    int left = 0;
    int right = arr.size() - 1;
    
    while (left <= right) {
        int mid = left + (right - left) / 2;
        emit("SEARCH_NARROW", "\"left\":" + to_string(left) + ", \"right\":" + to_string(right) + ", \"mid\":" + to_string(mid));
        emit("SEARCH_CHECK", "\"index\":" + to_string(mid) + ", \"value\":" + to_string(arr[mid]) + ", \"target\":" + to_string(target));
        
        if (arr[mid] == target) {
            emit("SEARCH_FOUND", "\"index\":" + to_string(mid) + ", \"value\":" + to_string(arr[mid]));
            return mid;
        } else if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    emit("SEARCH_NOT_FOUND", "\"target\":" + to_string(target));
    return -1;
}

int main() {
    vector<int> arr = {3, 9, 10, 27, 38, 43, 82};
    int target = 38;
    binary_search(arr, target);
    return 0;
}

File: src/algorithms/python/binary-search.py
import json


def emit(event_type, **kwargs):
    event = {"type": event_type}
    event.update(kwargs)
    print(json.dumps(event))


def binary_search(arr, target):
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = (left + right) // 2
        emit("SEARCH_NARROW", left=left, right=right, mid=mid)
        emit("SEARCH_CHECK", index=mid, value=arr[mid], target=target)
        if arr[mid] == target:
            emit("SEARCH_FOUND", index=mid, value=arr[mid])
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    emit("SEARCH_NOT_FOUND", target=target)
    return -1


if __name__ == "__main__":
    arr = [3, 9, 10, 27, 38, 43, 82]
    target = 38
    emit("INIT", array=arr[:])
    binary_search(arr, target)
